### PR TITLE
[#30184][runners][spark] GBK with global window translation OOM fix

### PR DESCRIPTION
**Pull request by mareksimunek**
More details in #30184

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Abacn** [comment]
Thanks for contribution. for the Spotless error, could run `./gradlew spotlessApply` to format the code and then push

File: runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java [hunk 1]
```diff
@@ -159,91 +160,100 @@ private static <K, V> Partitioner getPartitioner(
...
+                    if (consumed) {
+                      throw new IllegalStateException(
+                          "ValueIterator can't be iterated more than once otherwise there could be data lost");
+                    }
+                    consumed = true;
+                    return new AbstractIterator<V>() {
+
+                      @Override
+                      public V computeNext() {
+                        if (inner.hasNext() && inner.peek()._1.equals(currentKey)) {
```
Comment (je-ik) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/translation/GroupNonMergingWindowsFunctions.java:213]: This implies that Key must use deterministic Coder, otherwise this might not work. Is there any check that the Coder is deterministic before resorting to this optimization?

**Review by je-ik — COMMENTED** [review]
I have a few questions, please see the review. We also need to run VR tests for spark, please create empty file in path `.github/trigger_files/beam_PostCommit_Java_ValidatesRunner_Spark.json`, which will trigger the test for this PR. The file needs to be removed before merge.

**Comment by mareksimunek** [comment]
waiting on author
